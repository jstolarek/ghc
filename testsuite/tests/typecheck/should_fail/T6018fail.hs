{-# LANGUAGE TypeFamilies, DataKinds, UndecidableInstances #-}

module T6018fail where

import T6018Afail -- defines G, identical to F
import T6018Cfail -- imports H from T6018Bfail, defines some equations for H
import T6018Dfail -- imports H from T6018Bfail, defines conflicting eqns

type family F a b c = (result :: *) | result -> a b c
type instance F Int  Char Bool = Bool
type instance F Char Bool Int  = Int
type instance F Bool Int  Char = Int

type instance G Bool Int  Char = Int

type family I a b c = r | r -> a b
type instance I Int  Char Bool = Bool
type instance I Int  Int  Int  = Bool
type instance I Bool Int  Int  = Int

-- Id is injective...
type family Id a = result | result -> a
type instance Id a = a

-- ...but despite that we disallow a call to Id
type family IdProxy a = r | r -> a
type instance IdProxy a = Id a

data N = Z | S N

-- P is not injective, although the user declares otherwise. This
-- should be rejected on the grounds of calling a type family in the
-- RHS.
type family P (a :: N) (b :: N) = (r :: N) | r -> a b
type instance P  Z    m = m
type instance P (S n) m = S (P n m)

-- this is not injective - not all injective type variables mentioned
-- on LHS are mentioned on RHS
type family J a b c = r | r -> a b
type instance J Int b c = Char

-- same as above, but tyvar is now nested inside a tycon
type family K (a :: N) (b :: N) = (r :: N) | r -> a b
type instance K (S n) m = S m

type MaybeSyn a = Id a
type family L a = r | r -> a
type instance L a = MaybeSyn a
