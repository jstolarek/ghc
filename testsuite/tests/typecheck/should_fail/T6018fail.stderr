[1 of 5] Compiling T6018Bfail       ( T6018Bfail.hs, T6018Bfail.o )
[2 of 5] Compiling T6018Dfail       ( T6018Dfail.hs, T6018Dfail.o )
[3 of 5] Compiling T6018Cfail       ( T6018Cfail.hs, T6018Cfail.o )
[4 of 5] Compiling T6018Afail       ( T6018Afail.hs, T6018Afail.o )
[5 of 5] Compiling T6018fail        ( T6018fail.hs, T6018fail.o )

T6018Afail.hs:7:15:
    Family instance declarations violate injectivity declaration:
      G Char Bool Int -- Defined at T6018Afail.hs:7:15
      G Bool Int Char -- Defined at T6018fail.hs:14:15

T6018Cfail.hs:8:15:
    Family instance declarations violate injectivity declaration:
      T6018Bfail.H Char Bool Int -- Defined at T6018Cfail.hs:8:15
      T6018Bfail.H Bool Int Char -- Defined at T6018Dfail.hs:7:15

T6018fail.hs:11:15:
    Family instance declarations violate injectivity declaration:
      F Char Bool Int -- Defined at T6018fail.hs:11:15
      F Bool Int Char -- Defined at T6018fail.hs:12:15

T6018fail.hs:17:15:
    Family instance declarations violate injectivity declaration:
      I Int Char Bool -- Defined at T6018fail.hs:17:15
      I Int Int Int -- Defined at T6018fail.hs:18:15

T6018fail.hs:27:15:
    Calling type family `Id' is not allowed in injective type family equation:
      IdProxy a -- Defined at T6018fail.hs:27:15

T6018fail.hs:36:15:
    Calling type family `P' is not allowed in injective type family equation:
      P ('S n) m -- Defined at T6018fail.hs:36:15

T6018fail.hs:41:15:
    Family instance declaration violates injectivity declaration. Type variable `b' should appear in the type family equation:
      J Int b c -- Defined at T6018fail.hs:41:15

T6018fail.hs:45:15:
    Family instance declaration violates injectivity declaration. Type variable `n' should appear in the type family equation:
      K ('S n) m -- Defined at T6018fail.hs:45:15
